ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs

```diff
@@ -171,105 +216,321 @@ data InitLog blk =
+    either
+      (error . ("invariant violation: invalid current chain:" <>) . show)
+      (\(ledgerDbCheckpoints, w) -> return (LedgerDB{..}, w))
+    =<<
+      runExceptT (streamUpTo streamAPI from to InitFailureTooRecent (old, 0) push)
+  where
+    push :: blk
+         ->   (OldLedgerDB (ExtLedgerState blk), Word64)
+         -> m (OldLedgerDB (ExtLedgerState blk), Word64)
+    push blk !(!db, !replayed) = do
+      let ls =   forgetLedgerStateTracking
+               $ tickThenReapply (ledgerDbCfg cfg) blk
+               $ either unCheckpoint unCheckpoint . AS.head
+               $ db
+
+          !db' = AS.anchorNewest (maxRollbacks $ ledgerDbCfgSecParam cfg) (db AS.:> Checkpoint ls)
+
+          !replayed' = replayed + 1
+
+      return (db', replayed')
+
+-- | The new ledger DB is behind the old ledger DB.
+--
+-- Pushes blocks in the ImmutableDB up to the tip of the old ledger DB so that
+-- they end up being in sync.
+bringUpNewLedgerDB :: forall m blk. (Monad m, LedgerSupportsProtocol blk, InspectLedger blk)
+                   => Bool                                                                 -- ^ Whether to run both databases TODO: Still not implemented
+                   -> Tracer m (ReplayStart blk -> ReplayGoal blk -> TraceReplayEvent blk) -- ^ The replay tracer to emit events when replaying blocks
+                   -> LedgerDbCfg (ExtLedgerState blk)                                     -- ^ The LedgerDB configuration
+                   -> StreamAPI m blk                                                      -- ^ An API to stream blocks from the Immutable DB
+                   -> OnDiskLedgerStDb m (ExtLedgerState blk) blk                          -- ^ The backend handle
+                   -> OldLedgerDB (ExtLedgerState blk)                                     -- ^ The old-style database
+                   -> NewLedgerDB (ExtLedgerState blk)                                     -- ^ The new-style database that we want to bring up
+                   -> Point blk                                                            -- ^ At which point the old-style database is
+                   -> Point blk                                                            -- ^ At which point the new-style database is
+                   -> m (LedgerDB' blk, Word64)
+bringUpNewLedgerDB runDual tracer cfg streamAPI onDiskLedgerDbSt ledgerDbCheckpoints new to from =
+    either
+      (error . ("invariant violation: invalid current chain:" <>) . show)
+      (\(ledgerDbChangelog, w) -> return (LedgerDB{..}, w))
+    =<<
+      runExceptT (streamUpTo streamAPI from to InitFailureTooRecent (new, 0) push)
+  where
+    push :: blk
+         ->   (NewLedgerDB (ExtLedgerState blk), Word64)
+         -> m (NewLedgerDB (ExtLedgerState blk), Word64)
+    push blk !(db, replayed) = do
+      ls <- defaultReadKeySets (readKeySets onDiskLedgerDbSt) (withBlockReadSets blk db $ \lh -> return $ tickThenReapply (ledgerDbCfg cfg) blk lh)
```

Comment:
Can we discuss this use of `defaultReadKeySets`?

Reply:
This was just a copy-paste from `addBlockNew`, but it is gone now